examples/fool-patchguard.c

```diff
@@ -53,14 +146,118 @@ event_response_t read_callback(vmi_instance_t vmi, vmi_event_t *event)
 
-    printf("%s: at 0x%"PRIx64", on frame 0x%"PRIx64", permissions: %s\n",
-           __func__, event->x86_regs->rip, event->mem_event.gfn, str_access);
+    // this check is useless since we configured the callback only on a specific GFN
+    // however, if the read event was generic, we should filter early on the GFN we are expecting to intercept
+    if (!(event->mem_event.gfn == cb_data->target_gfn)) {
+        // not in the gfn we are looking for
+        return rsp;
+    }
+
+    if (!(event->mem_event.out_access & VMI_MEMACCESS_R)) {
+        // not a read event. skip.
+        return rsp;
+    }
+
+    // read a buffer of an x86 insn max size at RIP (15 Bytes)
+    uint8_t insn_buffer[MAX_SIZE_X86_INSN] = {0};
+    size_t bytes_read = 0;
+    if (VMI_FAILURE == vmi_read_va(vmi, event->x86_regs->rip, 0, MAX_SIZE_X86_INSN, insn_buffer, &bytes_read)) {
+        fprintf(stderr, "Failed to read buffer at RIP\n");
+        return rsp;
+    }
+
+    // check bytes_read
+    if (bytes_read != MAX_SIZE_X86_INSN) {
+        fprintf(stderr, "Failed to read enough bytes at RIP\n");
+        return rsp;
+    }
+
+    // disassemble next instruction with libbdisasm
+    uint8_t defcode = ND_CODE_32;
+    uint8_t defdata = ND_DATA_32;
+    if (cb_data->is64) {
+        defcode = ND_CODE_64;
+        defdata = ND_DATA_64;
+    }
+
+    INSTRUX rip_insn;
+    NDSTATUS status = NdDecodeEx(&rip_insn, insn_buffer, sizeof(insn_buffer), defcode, defdata);
+    if (!ND_SUCCESS(status)) {
+        fprintf(stderr, "Failed to decode instruction with libbdisasm: %x\n", status);
+        return rsp;
+    }
+
+    // determine memory access size
+    size_t access_size = 0;
+    if (!mem_access_size_from_insn(&rip_insn, &access_size)) {
+        return rsp;
+    }
+    // Debug
+    // printf("Read access size: %ld\n", access_size);
+
+    // find read guest physical addr
+    addr_t read_paddr = 0;
+    if (VMI_FAILURE == vmi_translate_kv2p(vmi, event->mem_event.gla, &read_paddr)) {
+        fprintf(stderr, "Failed to translate read virtual address\n");
+        return rsp;
+    }
+
+    // check if zones are overlapping
+    range_t read_zone = {
+        .start = read_paddr,
+        .size = access_size,
+        .end = read_paddr + access_size
+    };
+    range_t syscall_entry_zone = {
+        .start = cb_data->target_entry_paddr,
+        .size = KISERVICE_ENTRY_SIZE,
+        .end = cb_data->target_entry_paddr + KISERVICE_ENTRY_SIZE
+    };
+    range_t overlap = {0};
+    if (is_zone_read(&read_zone, &syscall_entry_zone, &overlap)) {
+        // overlap !
+        // get insn string
+        char insn_str[ND_MIN_BUF_SIZE];
+        NdToText(&rip_insn, 0, sizeof(insn_str), insn_str);
+        printf("Read on KiServiceTable %s entry (size: %ld)\n", cb_data->target, overlap.size);
+        printf("\t0x%"PRIx64": %s\n", event->x86_regs->rip, insn_str);
+
+        memset(&g_emul_read, 0, sizeof(g_emul_read));
+        g_emul_read.dont_free = true;
```

Comment:
@tklengyel I was setting the `dont_free` attribute here, as I reset the struct before.
I will change this to `1` since it's an `int`

Reply:
If a variable is a static global then you don't need to initialize it to 0, it's already initialized to 0.